src/encoded/static/components/matrix_encore.js

```diff
@@ -0,0 +1,619 @@
+            <div className="matrix-description">
+                <div className="matrix-description__text">
+                    The ENCORE project aims to study protein-RNA interactions by creating a map of RNA binding proteins (RBPs) encoded in the human genome and identifying the RNA elements that the RBPs bind to.
+                </div>
+            </div>
+        </div>
+        <div className="matrix-header__controls">
+            <div className="matrix-header__target-filter-controls">
+                <TargetFilter filterText={targetFilterText} textChangeHandler={textChangeHandler} />
+            </div>
+        </div>
+    </div>
+);
+
+MatrixHeader.propTypes = {
+    /** Matrix search result object */
+    context: PropTypes.object.isRequired,
+    /** Current value of filtering text */
+    targetFilterText: PropTypes.string,
+    /** Called when the user changes a search input field */
+    textChangeHandler: PropTypes.func.isRequired,
+};
+
+MatrixHeader.defaultProps = {
+    targetFilterText: '',
+};
+
+
+/**
+ * Custom react hook to handle matrix scrolling.
+ * @param {node} Ref to scrolling matrix element
+ *
+ * @return {array} States and actions usable to this hook's clients
+ */
+const useMatrixScrollHandler = (scrollElement) => {
+    const [scrolledRight, setScrolledRight] = React.useState(true);
+
+    /**
+     * Show a scroll indicator depending on current scrolled position.
+     * @param {object} element DOM element to apply shading to
+     */
+    const handleScrollIndicator = React.useCallback((element) => {
+        // Have to use a "roughly equal to" test because of an MS Edge bug mentioned here:
+        // https://stackoverflow.com/questions/30900154/workaround-for-issue-with-ie-scrollwidth
+        const scrollDiff = Math.abs((element.scrollWidth - element.scrollLeft) - element.clientWidth);
+        if (scrollDiff < 2 && !scrolledRight) {
+            // Right edge of matrix scrolled into view.
+            setScrolledRight(true);
+        } else if (scrollDiff >= 2 && scrolledRight) {
+            // Right edge of matrix scrolled out of view.
+            setScrolledRight(false);
+        }
+    }, [scrolledRight]);
+
+    React.useEffect(() => {
+        if (scrollElement.current) {
+            // Install the scroll and resize event handlers.
+            const handleScrollEvent = event => handleScrollIndicator(event.target);
+            const handleResizeEvent = () => handleScrollIndicator(scrollElement.current);
+
+            // Cache the reference to the scrollable matrix <div> so that we can remove the "scroll"
+            // event handler on unmount, when scrollElement.current might no longer point at this <div>.
+            const matrixNode = scrollElement.current;
+
+            // Attach the scroll- and resize- event handlers, then force the initial calculation of
+            // `scrolledRight`.
+            scrollElement.current.addEventListener('scroll', handleScrollEvent);
```

Comment:
In theory, this hook can be called multiple times without the unmounting part called. If that theory is correct. Do you detach the previously binded scroll and resize event somewhere else too? Otherwise, that may be a memory leak.

Reply:
The unmounting callback gets called before useEffect gets called after the first time. So the order of execution becomes:

addEventListener
... _time passes_
removeEventListener
addEventListener
... _time passes_
removeEventListener
addEventListener
... _and so on_

https://reactjs.org/docs/hooks-reference.html#cleaning-up-an-effect